Windows build

Currently the build fails on Windows.

```
 Compiling beamup v0.1.0 (D:\a\beamup\beamup)
error[E0433]: failed to resolve: use of undeclared crate or module `windows_sys`
  --> src\run.rs:31:9
   |
31 |     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
   |         ^^^^^^^^^^^ use of undeclared crate or module `windows_sys`

error[E0433]: failed to resolve: use of undeclared crate or module `windows_sys`
  --> src\run.rs:32:9
   |
32 |     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
   |         ^^^^^^^^^^^ use of undeclared crate or module `windows_sys`

error[E0433]: failed to resolve: use of undeclared crate or module `io`
  --> src\run.rs:40:24
   |
40 |             return Err(io::Error::new(
   |                        ^^ use of undeclared crate or module `io`
   |
help: consider importing one of these items
   |
1  + use clap::Error;
   |
1  + use clap::error::Error;
   |
1  + use core::error::Error;
   |
1  + use core::fmt::Error;
   |
     and 13 other candidates
help: if you import `Error`, refer to it directly
   |
40 -             return Err(io::Error::new(
40 +             return Err(Error::new(
   |

error[E0433]: failed to resolve: use of undeclared crate or module `io`
  --> src\run.rs:41:17
   |
41 |                 io::ErrorKind::Other,
   |                 ^^ use of undeclared crate or module `io`
   |
help: consider importing one of these items
   |
1  + use clap::error::ErrorKind;
   |
1  + use std::io::ErrorKind;
   |
1  + use tokio::io::ErrorKind;
   |
1  + use ureq::ErrorKind;
   |
help: if you import `ErrorKind`, refer to it directly
   |
41 -                 io::ErrorKind::Other,
41 +                 ErrorKind::Other,
   |

error[E0308]: mismatched types
  --> src\run.rs:47:5
   |
30 | fn exec(cmd: &mut Command) -> Result<(), Report> {
   |                               ------------------ expected `Result<(), ErrReport>` because of return type
...
47 |     cmd.status()
   |     ^^^^^^^^^^^^ expected `Result<(), Report>`, found `Result<ExitStatus, Error>`
   |
   = note: expected enum `Result<(), ErrReport>`
              found enum `Result<ExitStatus, std::io::Error>`
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,12 +1,19 @@
 [package]
 name = "beamup"
 version = "0.2.0"
+description = " Install and control active BEAM languages and components"
 edition = "2021"
 authors = ["Tristan Sloughter <tristan@sloughter.dev>"]
 repository = "https://github.com/tsloughter/beamup"
 license = "Apache-2.0"
 readme = "README.md"
 homepage = "https://github.com/tsloughter/beamup"
+
+[package.metadata.wix]
+upgrade-guid = "67FB3E2E-2204-4262-B187-C940B9E293E1"
+path-guid = "7600FDEA-F157-41B8-A433-A81FC6A057DF"
+license = false
+eula = false
 
 [dependencies]
 clap = {version = "4.5.8", features = ["derive", "color"]}
@@ -30,6 +37,7 @@ flate2 = "1.0.30"
 color-eyre = "0.6.2"
 http = "1.1.0"
 ureq = "2.10.0"
+windows-sys = "0.59.0"
 
 # The profile that 'cargo dist' will build with
 [profile.dist]
@@ -43,8 +51,12 @@ cargo-dist-version = "0.19.1"
 # CI backends to support
 ci = "github"
 # The installers to generate for each app
-installers = ["shell"]
+installers = ["shell", "powershell", "msi"]
 # Target platforms to build apps for (Rust target-triple syntax)
-targets = ["aarch64-apple-darwin", "x86_64-apple-darwin", "x86_64-unknown-linux-gnu"]
+targets = ["aarch64-apple-darwin", "x86_64-apple-darwin", "x86_64-unknown-linux-gnu", "x86_64-unknown-linux-musl", "x86_64-pc-windows-msvc"]
 # Publish jobs to run in CI
 pr-run-mode = "plan"
+# Path that installers should place binaries in
+install-path = "CARGO_HOME"
+# Whether to install an updater program
+install-updater = false

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -37,7 +37,8 @@ flate2 = "1.0.30"
 color-eyre = "0.6.2"
 http = "1.1.0"
 ureq = "2.10.0"
-windows-sys = "0.59.0"
+windows-sys = { version = "0.59.0", features = ["Win32_System", "Win32_System_Console"] }
+zip = "2.2.0"
 
 # The profile that 'cargo dist' will build with
 [profile.dist]

--- a/README.md
+++ b/README.md
@@ -3,9 +3,9 @@
 [![Release](https://github.com/tsloughter/beamup/actions/workflows/release.yml/badge.svg)](https://github.com/tsloughter/beamup/actions/workflows/release.yml)
 
 A tool for installing languages that run on the [Erlang
-VM](https://www.erlang.org/) (BEAM) on related components.
-As of now only Erlang and [Gleam](https://gleam.run/) are supported, with [Elixir](http://elixir-lang.org/)
-to come next.
+VM](https://www.erlang.org/) (BEAM) and related components. As of now only
+Erlang and [Gleam](https://gleam.run/) are supported, with
+[Elixir](http://elixir-lang.org/) to come next.
 
 ## Install
 
@@ -21,16 +21,30 @@ using [cargo](https://doc.rust-lang.org/cargo/).
 
 ## Usage
 
-`beamup` will store configuration at `~/.config/beamup/config.toml` by default
-(no override is yet supported) and local configuration at `./.beamup.toml`.
+`beamup` will store configuration at:
 
-In `~/.local/bin/` there are binaries created as hard links to the `beamup`
-executable for each language command, i.e. `gleam`, `erlc`, `erl`, etc.
+- Linux: `~/.config/beamup/config.toml` 
+- Mac: `~/Library/Application Support/beamup/config.toml`
+- Windows: `~\AppData\Local\beamup\config.toml`
 
-Installs are currently done to `~/.cache/beamup/<language>/<id>`. I don't like
-that. It doesn't make sense to be under `.cache`. It may be moved under
-`~/.local` as well in the next release. Or somewhere else more appropriate by
-XDG standards.
+Local configuration to set a language/component to use in a specific directory
+is in `./.beamup.toml`.
+
+Hard links to the `beamup` executable for each language command, i.e. `gleam`,
+`erlc`, `erl`, etc, is created in the following directory:
+
+- Linux: `$XDG_BIN_HOME` or `$XDG_DATA_HOME/../bin` or `$HOME/.local/bin`
+- Mac: `~/.beamup/bin`
+- Windows: `~\.beamup\bin`
+
+This directory must be added to your `PATH` for `beamup` installs to work.
+
+Installs are currently done to the applications data directory. This defaults
+to:
+
+- Linux: `~/.local/share/beamup/<language>/<id>`
+- Mac: `~/Library/Application Support/beamup/<language>/<id>`
+- Windows: `~\AppData\Local\beamup\<language>\<id>`
 
 For languages that support building from source you can pass additional build
 options (like what is passed to `./configure` for Erlang) with either the

--- a/src/cmd/build.rs
+++ b/src/cmd/build.rs
@@ -48,10 +48,12 @@ pub fn run(
     git_ref: &GitRef,
     id: &String,
     _repo: &Option<String>,
-    _force: &Option<bool>,
+    force: &Option<bool>,
     config: &config::Config,
 ) -> Result<String> {
     debug!("Building {language} from source from git ref={git_ref} with id={id}");
+
+    let release_dir = config::language_release_dir(language, id, force)?;
 
     //maybe grab configure options from environment
     let key = "BEAMUP_BUILD_OPTIONS";
@@ -73,7 +75,6 @@ pub fn run(
         )
     })?;
 
-    let release_dir = config::language_release_dir(language.to_owned(), id.to_owned());
     debug!("unpacking source tarball {tar_gz:?} to {out_dir:?}");
     let tar = GzDecoder::new(tar_gz);
     let mut archive = Archive::new(tar);

--- a/src/cmd/install.rs
+++ b/src/cmd/install.rs
@@ -6,6 +6,7 @@ use flate2::read::GzDecoder;
 use std::fs::File;
 use tar::Archive;
 use tempdir::TempDir;
+use zip;
 
 pub fn run(
     language: &languages::Language,
@@ -17,17 +18,28 @@ pub fn run(
 ) -> Result<String, Report> {
     let out_dir = TempDir::new(github_repo.repo.as_str())?;
     let file = download_asset(language, out_dir.path(), github_repo, release)?;
-
-    let tar_gz = File::open(&file).wrap_err_with(|| {
+    let release_dir = config::language_release_dir(language.to_owned(), id.to_owned());
+    debug!("file {:?} downloaded", file);
+    let open_file = File::open(&file).wrap_err_with(|| {
         format!(
             "Downloaded Github asset for release {} into file {:?} not found",
             release, &file
         )
     })?;
-    let tar = GzDecoder::new(tar_gz);
-    let mut archive = Archive::new(tar);
-    let release_dir = config::language_release_dir(language.to_owned(), id.to_owned());
-    archive.unpack(&release_dir.join("bin"))?;
+
+    // TODO: better ways to check the type than the extension
+    let ext = file.extension().map_or("", |e| e.to_str().unwrap_or(""));
+    match ext {
+        "zip" => {
+            let mut archive = zip::ZipArchive::new(open_file)?; 
+            archive.extract(&release_dir.join("bin"))?;
+        },
+        _ => {
+            let tar = GzDecoder::new(open_file);
+            let mut archive = Archive::new(tar);
+            archive.unpack(&release_dir.join("bin"))?;
+    }
+};
 
     Ok(release_dir.into_os_string().into_string().unwrap())
 }

--- a/src/cmd/install.rs
+++ b/src/cmd/install.rs
@@ -4,11 +4,10 @@ use crate::languages;
 use color_eyre::{eyre::eyre, eyre::Report, eyre::Result, eyre::WrapErr};
 use flate2::read::GzDecoder;
 use std::fs::File;
-use std::process::{Command, ExitStatus};
+use std::path::PathBuf;
 use tar::Archive;
 use tempdir::TempDir;
 use zip;
-use std::path::PathBuf;
 
 pub fn run(
     language: &languages::Language,
@@ -64,9 +63,9 @@ fn exe_run(_cmd: PathBuf, _release_dir: String) -> Result<(), Report> {
 // thanks rustup command.rs
 #[cfg(windows)]
 fn exe_run(file: PathBuf, release_dir: String) -> Result<ExitStatus, Report> {
+    use std::os::windows::process::CommandExt;
     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
-    use std::os::windows::process::CommandExt;
 
     unsafe extern "system" fn ctrlc_handler(_: u32) -> BOOL {
         // Do nothing. Let the child process handle it.

--- a/src/cmd/install.rs
+++ b/src/cmd/install.rs
@@ -9,6 +9,9 @@ use tar::Archive;
 use tempdir::TempDir;
 use zip;
 
+#[cfg(windows)]
+use std::process::ExitStatus;
+
 pub fn run(
     language: &languages::Language,
     github_repo: &GithubRepo,
@@ -34,9 +37,6 @@ pub fn run(
     match ext {
         "exe" => {
             let release_dir = release_dir.into_os_string().into_string().unwrap();
-            // let file = file.into_os_string().into_string().unwrap();
-            // let cmd = format!("{file:}");
-
             exe_run(file, release_dir.clone())?;
             Ok(release_dir)
         }
@@ -65,6 +65,7 @@ fn exe_run(_cmd: PathBuf, _release_dir: String) -> Result<(), Report> {
 #[cfg(windows)]
 fn exe_run(file: PathBuf, release_dir: String) -> Result<ExitStatus, Report> {
     use std::os::windows::process::CommandExt;
+    use std::process::Command;
     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
 

--- a/src/cmd/install.rs
+++ b/src/cmd/install.rs
@@ -1,12 +1,14 @@
 use crate::config;
 use crate::github::{download_asset, GithubRepo};
 use crate::languages;
-use color_eyre::{eyre::Report, eyre::Result, eyre::WrapErr};
+use color_eyre::{eyre::eyre, eyre::Report, eyre::Result, eyre::WrapErr};
 use flate2::read::GzDecoder;
 use std::fs::File;
+use std::process::{Command, ExitStatus};
 use tar::Archive;
 use tempdir::TempDir;
 use zip;
+use std::path::PathBuf;
 
 pub fn run(
     language: &languages::Language,
@@ -30,16 +32,55 @@ pub fn run(
     // TODO: better ways to check the type than the extension
     let ext = file.extension().map_or("", |e| e.to_str().unwrap_or(""));
     match ext {
+        "exe" => {
+            let release_dir = release_dir.into_os_string().into_string().unwrap();
+            // let file = file.into_os_string().into_string().unwrap();
+            // let cmd = format!("{file:}");
+
+            exe_run(file, release_dir.clone())?;
+            Ok(release_dir)
+        }
         "zip" => {
-            let mut archive = zip::ZipArchive::new(open_file)?; 
+            let mut archive = zip::ZipArchive::new(open_file)?;
             archive.extract(&release_dir.join("bin"))?;
-        },
+            Ok(release_dir.into_os_string().into_string().unwrap())
+        }
         _ => {
             let tar = GzDecoder::new(open_file);
             let mut archive = Archive::new(tar);
             archive.unpack(&release_dir.join("bin"))?;
+            Ok(release_dir.into_os_string().into_string().unwrap())
+        }
     }
-};
+}
 
-    Ok(release_dir.into_os_string().into_string().unwrap())
+#[cfg(unix)]
+fn exe_run(_cmd: PathBuf, _release_dir: String) -> Result<(), Report> {
+    Err(eyre!(
+        "Attempted to execute a Windows exeutable on a non-Windows system"
+    ))
+}
+
+// thanks rustup command.rs
+#[cfg(windows)]
+fn exe_run(file: PathBuf, release_dir: String) -> Result<ExitStatus, Report> {
+    use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
+    use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
+    use std::os::windows::process::CommandExt;
+
+    unsafe extern "system" fn ctrlc_handler(_: u32) -> BOOL {
+        // Do nothing. Let the child process handle it.
+        TRUE
+    }
+    unsafe {
+        if SetConsoleCtrlHandler(Some(ctrlc_handler), TRUE) == FALSE {
+            return Err(eyre!("Unable to set console handler",));
+        }
+    }
+
+    let mut binding = Command::new(file);
+    let cmd = binding.raw_arg(&format!("/S /D={release_dir:}"));
+    debug!("Command being run: {cmd:?}");
+
+    Ok(cmd.status()?)
 }

--- a/src/cmd/install.rs
+++ b/src/cmd/install.rs
@@ -15,11 +15,12 @@ pub fn run(
     release: &str,
     id: &String,
     _repo: &Option<String>,
-    _force: &Option<bool>,
+    force: &Option<bool>,
 ) -> Result<String, Report> {
+    let release_dir = config::language_release_dir(language, id, force)?;
+
     let out_dir = TempDir::new(github_repo.repo.as_str())?;
     let file = download_asset(language, out_dir.path(), github_repo, release)?;
-    let release_dir = config::language_release_dir(language.to_owned(), id.to_owned());
     debug!("file {:?} downloaded", file);
     let open_file = File::open(&file).wrap_err_with(|| {
         format!(

--- a/src/cmd/update_links.rs
+++ b/src/cmd/update_links.rs
@@ -1,11 +1,11 @@
+use crate::config;
 use crate::languages;
 use std::path::Path;
 
 pub fn run(maybe_language: Option<&languages::Language>) {
     for (b, l) in languages::BIN_MAP.iter() {
         if maybe_language.is_none() || maybe_language.is_some_and(|x| *x == *l) {
-            let home_dir = dirs::home_dir().unwrap(); //config::lookup_cache_dir(&config);
-            let bin_dir = Path::new(&home_dir).join(".local").join("bin");
+            let bin_dir = config::bin_dir();
             let _ = std::fs::create_dir_all(&bin_dir);
 
             let link = Path::new(&bin_dir).join(b);

--- a/src/cmd/update_links.rs
+++ b/src/cmd/update_links.rs
@@ -1,8 +1,10 @@
 use crate::config;
 use crate::languages;
+use color_eyre::{eyre::eyre, eyre::Result};
 use std::path::Path;
 
-pub fn run(maybe_language: Option<&languages::Language>) {
+pub fn run(maybe_language: Option<&languages::Language>) -> Result<()> {
+    let mut has_err = false;
     for (b, l) in languages::BIN_MAP.iter() {
         if maybe_language.is_none() || maybe_language.is_some_and(|x| *x == *l) {
             let bin_dir = config::bin_dir();
@@ -12,17 +14,20 @@ pub fn run(maybe_language: Option<&languages::Language>) {
             let beamup_exe = std::env::current_exe().unwrap();
             debug!("linking {:?} to {:?}", link, beamup_exe);
             let _ = std::fs::remove_file(&link);
-            // Ok(()) =>
             match std::fs::hard_link(beamup_exe, &link) {
                 Ok(()) => {}
                 Err(e) => {
-                    error!("Failed to symlink {:?}: {}", link, e);
+                    has_err = true;
+                    error!("Failed to link {:?}: {}", link, e);
                 }
             }
-            // Err(e) => {
-            //         error!("Failed to remove symlink {:?}: {}", link, e);
-            //     }
-            // }
         }
+    }
+
+    // TODO: should do a multi-report error instead of this
+    if has_err {
+        Err(eyre!("Some links failed to be created"))
+    } else {
+        Ok(())
     }
 }

--- a/src/config.rs
+++ b/src/config.rs
@@ -204,8 +204,8 @@ pub fn language_release_dir(
     id: &String,
     force: &Option<bool>,
 ) -> Result<PathBuf> {
-    let cache_dir = dirs::data_local_dir();
-    let release_dir = cache_dir
+    let data_dir = dirs::data_local_dir();
+    let release_dir = data_dir
         .unwrap()
         .join("beamup")
         .join(language.to_string())
@@ -246,20 +246,20 @@ pub fn home_config_file() -> Result<String> {
         Some(d) => d,
         None => return Err(eyre!("no home directory available")),
     };
-    let cache_dir = match dirs::data_local_dir() {
+    let data_dir = match dirs::data_local_dir() {
         Some(d) => d,
         None => return Err(eyre!("no home directory available")),
     };
 
     let default_config = config_dir.join("beamup").join(CONFIG_FILE);
-    let default_cache = cache_dir.join("beamup");
+    let default_data = data_dir.join("beamup");
 
     let _ = fs::create_dir_all(config_dir.join("beamup"));
-    let _ = fs::create_dir_all(cache_dir.join("beamup"));
+    let _ = fs::create_dir_all(data_dir.join("beamup"));
 
     if !default_config.exists() {
         let config = Config {
-            install_dir: default_cache.to_str().unwrap().to_string(),
+            install_dir: default_data.to_str().unwrap().to_string(),
             erlang: Some(LanguageConfig {
                 default: None,
                 installs: toml::Table::new(),

--- a/src/config.rs
+++ b/src/config.rs
@@ -199,7 +199,11 @@ pub fn add_install(
     let _ = write_config(config_file, new_config);
 }
 
-pub fn language_release_dir(language: languages::Language, id: String) -> PathBuf {
+pub fn language_release_dir(
+    language: &languages::Language,
+    id: &String,
+    force: &Option<bool>,
+) -> Result<PathBuf> {
     let cache_dir = dirs::data_local_dir();
     let release_dir = cache_dir
         .unwrap()
@@ -207,9 +211,24 @@ pub fn language_release_dir(language: languages::Language, id: String) -> PathBu
         .join(language.to_string())
         .join(id);
 
+    match release_dir.try_exists() {
+        Ok(true) =>
+            match force {
+                Some(true) => {
+                    info!("Force enabled. Deleting existing release directory {release_dir:?}");
+                    fs::remove_dir_all(&release_dir)?
+                },
+                _ => return Err(eyre!("Release directory for id {id:} already exists. Use `-f` to delete {release_dir:?} and recreate instead of giving this error.")),
+            }
+        Ok(false) => {},
+        Err(e) => return Err(eyre!(
+            "Unable to check for existence of release directory for id {id}: {e:?}"
+        )),
+    };
+
     let _ = std::fs::create_dir_all(&release_dir);
 
-    release_dir
+    Ok(release_dir)
 }
 
 pub fn bin_dir() -> PathBuf {

--- a/src/config.rs
+++ b/src/config.rs
@@ -200,7 +200,7 @@ pub fn add_install(
 }
 
 pub fn language_release_dir(language: languages::Language, id: String) -> PathBuf {
-    let cache_dir = dirs::cache_dir();
+    let cache_dir = dirs::data_local_dir();
     let release_dir = cache_dir
         .unwrap()
         .join("beamup")
@@ -212,12 +212,22 @@ pub fn language_release_dir(language: languages::Language, id: String) -> PathBu
     release_dir
 }
 
+pub fn bin_dir() -> PathBuf {
+    match dirs::executable_dir() {
+        Some(bin_dir) => bin_dir,
+        None => {
+            let home_dir = dirs::home_dir().unwrap();
+            Path::new(&home_dir).join(".beamup").join("bin")
+        }
+    }
+}
+
 pub fn home_config_file() -> Result<String> {
-    let config_dir = match dirs::config_dir() {
+    let config_dir = match dirs::config_local_dir() {
         Some(d) => d,
         None => return Err(eyre!("no home directory available")),
     };
-    let cache_dir = match dirs::cache_dir() {
+    let cache_dir = match dirs::data_local_dir() {
         Some(d) => d,
         None => return Err(eyre!("no home directory available")),
     };

--- a/src/github.rs
+++ b/src/github.rs
@@ -21,25 +21,38 @@ pub struct GithubRepo {
 }
 
 fn asset_name(language: &Language, tag: &str) -> Result<String> {
-    let suffix = match language {
-        Language::Gleam => match (std::env::consts::ARCH, std::env::consts::OS) {
-            ("x86_64", "linux") => "x86_64-unknown-linux-musl.tar.gz",
-            ("aarch64", "linux") => "aarch64-unknown-linux-musl.tar.gz",
-            ("x86_64", "macos") => "x86_64-apple-darwin.tar.gz",
-            ("aarch64", "macos") => "aarch64-apple-darwin.tar.gz",
-            ("x86_64", "windows") => "x86_64-pc-windows-msvc.zip",
-            (arch, os) => {
-                let e: Report = eyre!("no {language} asset found to support arch:{arch} os:{os}");
-                return Err(e);
+    let asset_name = match language {
+        Language::Gleam => {
+            let suffix = match (std::env::consts::ARCH, std::env::consts::OS) {
+                ("x86_64", "linux") => "x86_64-unknown-linux-musl.tar.gz",
+                ("aarch64", "linux") => "aarch64-unknown-linux-musl.tar.gz",
+                ("x86_64", "macos") => "x86_64-apple-darwin.tar.gz",
+                ("aarch64", "macos") => "aarch64-apple-darwin.tar.gz",
+                ("x86_64", "windows") => "x86_64-pc-windows-msvc.zip",
+                (arch, os) => {
+                    let e: Report =
+                        eyre!("no {language} asset found to support arch:{arch} os:{os}");
+                    return Err(e);
+                }
+            };
+
+            format!("{language}-{tag}-{suffix}")
+        }
+        Language::Erlang => {
+            let vsn = tag.strip_prefix("OTP-").unwrap_or(tag);
+            match (std::env::consts::ARCH, std::env::consts::OS) {
+                ("x86", "windows") => format!("otp_win32_{vsn}.exe"),
+                ("x86_64", "windows") => format!("otp_win64_{vsn}.exe"),
+                (arch, os) => {
+                    let e: Report =
+                        eyre!("no {language} asset found to support arch:{arch} os:{os}");
+                    return Err(e);
+                }
             }
-        },
-        _ => {
-            let e: Report = eyre!("Github binary asset download for {language} not supported");
-            return Err(e);
         }
     };
 
-    Ok(format!("{language}-{tag}-{suffix}"))
+    Ok(asset_name)
 }
 
 pub fn print_releases(GithubRepo { org, repo }: &GithubRepo) {

--- a/src/github.rs
+++ b/src/github.rs
@@ -171,7 +171,7 @@ pub fn download_asset(
 
     match assets.iter().find(|&asset| *asset.name == asset_name) {
         Some(asset) => {
-            let file = out_dir.join(repo.to_owned() + ".tar.gz");
+            let file = out_dir.join(asset_name);
             let dest = std::fs::File::create(&file)
                 .wrap_err_with(|| format!("Failed to create asset download file {:?}", file))?;
 

--- a/src/languages.rs
+++ b/src/languages.rs
@@ -6,6 +6,7 @@ use color_eyre::{eyre::eyre, eyre::Result};
 pub mod erlang;
 pub mod gleam;
 
+#[cfg(unix)]
 pub const BIN_MAP: &[(&str, languages::Language)] = &[
     ("gleam", languages::Language::Gleam),
     ("ct_run", languages::Language::Erlang),
@@ -19,6 +20,22 @@ pub const BIN_MAP: &[(&str, languages::Language)] = &[
     ("run_test", languages::Language::Erlang),
     ("to_erl", languages::Language::Erlang),
     ("typer", languages::Language::Erlang),
+];
+
+#[cfg(windows)]
+pub const BIN_MAP: &[(&str, languages::Language)] = &[
+    ("gleam.exe", languages::Language::Gleam),
+    ("ct_run.exe", languages::Language::Erlang),
+    ("dialyzer.exe", languages::Language::Erlang),
+    ("epmd.exe", languages::Language::Erlang),
+    ("erl.exe", languages::Language::Erlang),
+    ("erlc.exe", languages::Language::Erlang),
+    ("erl_call.exe", languages::Language::Erlang),
+    ("escript.exe", languages::Language::Erlang),
+    ("run_erl.exe", languages::Language::Erlang),
+    ("run_test.exe", languages::Language::Erlang),
+    ("to_erl.exe", languages::Language::Erlang),
+    ("typer.exe", languages::Language::Erlang),
 ];
 
 #[derive(ValueEnum, Debug, Clone, PartialEq)]

--- a/src/main.rs
+++ b/src/main.rs
@@ -233,11 +233,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
                 language, release, id, repo, force
             );
 
-            if *language != languages::Language::Gleam {
-                return Err(eyre!(
-                    "install command not supported yet for language {language:?}"
-                ));
-            }
+            let _ = check_if_supported(language);
 
             // if no user supplied id then use the name of
             // the release to install
@@ -320,6 +316,23 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
         }
         _ => Err(eyre!("subcommand not implemented yet")),
     }
+}
+
+fn check_if_supported(language: &languages::Language) -> Result<()> {
+    if *language != languages::Language::Gleam && *language == languages::Language::Erlang {
+        // install command for Erlang only supports Windows on x86 or x86_64 at this time
+        match (std::env::consts::ARCH, std::env::consts::OS) {
+            ("x86", "windows") => return Ok(()),
+            ("x86_64", "windows") => return Ok(()),
+            (os, arch) => {
+                return Err(eyre!(
+                    "install command not supported yet for language {language:?} on {os:?} {arch:?}"
+                ))
+            }
+        }
+    }
+
+    Ok(())
 }
 
 fn setup_logging() {

--- a/src/main.rs
+++ b/src/main.rs
@@ -242,7 +242,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
             let github_repo = languages::get_github_repo(language);
 
             let dir = cmd::install::run(language, &github_repo, release, id, repo, force)?;
-            cmd::update_links::run(Some(language));
+            let _ = cmd::update_links::run(Some(language))?;
 
             config::add_install(language, id, dir, config_file, config);
 
@@ -251,7 +251,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
         SubCommands::UpdateLinks => {
             debug!("running update-links");
 
-            cmd::update_links::run(None);
+            let _ = cmd::update_links::run(None)?;
 
             info!("Updated links of language binaries to current beamup install");
 
@@ -313,7 +313,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
 
             info!("Building {:?} for ref={} id={}", language, git_ref, id);
 
-            cmd::update_links::run(Some(language));
+            let _ = cmd::update_links::run(Some(language))?;
             config::add_install(language, &id, dir, config_file, config);
 
             Ok(())

--- a/src/main.rs
+++ b/src/main.rs
@@ -233,7 +233,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
                 language, release, id, repo, force
             );
 
-            let _ = check_if_supported(language);
+            let _ = check_if_install_supported(language);
 
             // if no user supplied id then use the name of
             // the release to install
@@ -322,7 +322,8 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
     }
 }
 
-fn check_if_supported(language: &languages::Language) -> Result<()> {
+// only Gleam on any platform and Erlang on Windows support binary install at this time
+fn check_if_install_supported(language: &languages::Language) -> Result<()> {
     if *language != languages::Language::Gleam && *language == languages::Language::Erlang {
         // install command for Erlang only supports Windows on x86 or x86_64 at this time
         match (std::env::consts::ARCH, std::env::consts::OS) {

--- a/src/main.rs
+++ b/src/main.rs
@@ -293,6 +293,10 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
                 ));
             }
 
+            if std::env::consts::OS == "windows" {
+                return Err(eyre!("build command not supported yet for Windows"));
+            }
+
             let git_ref = match release {
                 None => match branch {
                     None => {

--- a/src/main.rs
+++ b/src/main.rs
@@ -353,7 +353,7 @@ fn main() -> Result<(), Report> {
     let binname = args.next().unwrap();
     let f = Path::new(&binname).file_name().unwrap();
 
-    if f.eq("beamup") {
+    if f.eq("beamup") || f.eq("beamup.exe") {
         match env::current_exe() {
             Ok(bin_path) => {
                 debug!("current bin path: {}", bin_path.display());

--- a/src/main.rs
+++ b/src/main.rs
@@ -218,6 +218,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
         // }
         SubCommands::Releases(ReleasesArgs { language, .. }) => {
             debug!("running releases: repo={:?}", language);
+            // TODO: should return Result type
             cmd::releases::run(language);
             Ok(())
         }
@@ -242,8 +243,9 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
             let github_repo = languages::get_github_repo(language);
 
             let dir = cmd::install::run(language, &github_repo, release, id, repo, force)?;
-            let _ = cmd::update_links::run(Some(language))?;
+            cmd::update_links::run(Some(language))?;
 
+            // TODO: should return Result type
             config::add_install(language, id, dir, config_file, config);
 
             Ok(())
@@ -251,7 +253,7 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
         SubCommands::UpdateLinks => {
             debug!("running update-links");
 
-            let _ = cmd::update_links::run(None)?;
+            cmd::update_links::run(None)?;
 
             info!("Updated links of language binaries to current beamup install");
 
@@ -313,7 +315,9 @@ fn handle_command(_bin_path: PathBuf) -> Result<(), Report> {
 
             info!("Building {:?} for ref={} id={}", language, git_ref, id);
 
-            let _ = cmd::update_links::run(Some(language))?;
+            cmd::update_links::run(Some(language))?;
+
+            // TODO: this should Result type
             config::add_install(language, &id, dir, config_file, config);
 
             Ok(())

--- a/src/run.rs
+++ b/src/run.rs
@@ -1,10 +1,11 @@
 use color_eyre::{eyre::Report, eyre::Result};
 use std::env::Args;
 
-use std::path::*;
-use std::process::{Command, ExitStatus};
-
 use crate::config;
+use std::path::*;
+use std::process::Command;
+#[cfg(windows)]
+use std::process::ExitStatus;
 
 pub fn run(bin: &str, args: Args) -> Result<(), Report> {
     // no -c argument available in this case

--- a/src/run.rs
+++ b/src/run.rs
@@ -2,7 +2,7 @@ use color_eyre::{eyre::Report, eyre::Result};
 use std::env::Args;
 
 use std::path::*;
-use std::process::Command;
+use std::process::{Command, ExitStatus};
 
 use crate::config;
 
@@ -27,7 +27,7 @@ fn exec(cmd: &mut Command) -> Result<(), Report> {
 
 // thanks rustup command.rs
 #[cfg(windows)]
-fn exec(cmd: &mut Command) -> Result<(), Report> {
+fn exec(cmd: &mut Command) -> Result<ExitStatus, Report> {
     use color_eyre::eyre::eyre;
     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
@@ -42,6 +42,5 @@ fn exec(cmd: &mut Command) -> Result<(), Report> {
         }
     }
 
-    let _ = cmd.status();
-    Ok(())
+    Ok(cmd.status()?)
 }

--- a/src/run.rs
+++ b/src/run.rs
@@ -28,6 +28,7 @@ fn exec(cmd: &mut Command) -> Result<(), Report> {
 // thanks rustup command.rs
 #[cfg(windows)]
 fn exec(cmd: &mut Command) -> Result<(), Report> {
+    use color_eyre::eyre::eyre;
     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
 
@@ -37,12 +38,10 @@ fn exec(cmd: &mut Command) -> Result<(), Report> {
     }
     unsafe {
         if SetConsoleCtrlHandler(Some(ctrlc_handler), TRUE) == FALSE {
-            return Err(io::Error::new(
-                io::ErrorKind::Other,
-                "Unable to set console handler",
-            ));
+            return Err(eyre!("Unable to set console handler",));
         }
     }
 
-    cmd.status()
+    let _ = cmd.status();
+    Ok(())
 }

--- a/src/run.rs
+++ b/src/run.rs
@@ -1,13 +1,11 @@
-use color_eyre::{eyre::Report, eyre::Result};
+use color_eyre::eyre::Result;
 use std::env::Args;
 
 use crate::config;
 use std::path::*;
 use std::process::Command;
-#[cfg(windows)]
-use std::process::ExitStatus;
 
-pub fn run(bin: &str, args: Args) -> Result<(), Report> {
+pub fn run(bin: &str, args: Args) -> Result<()> {
     // no -c argument available in this case
     let dir = config::install_to_use(bin)?;
     let cmd = Path::new(&dir).join("bin").join(bin);
@@ -21,14 +19,14 @@ pub fn run(bin: &str, args: Args) -> Result<(), Report> {
 }
 
 #[cfg(unix)]
-fn exec(cmd: &mut Command) -> Result<(), Report> {
+fn exec(cmd: &mut Command) -> Result<()> {
     use std::os::unix::prelude::*;
     Err(cmd.exec().into())
 }
 
 // thanks rustup command.rs
 #[cfg(windows)]
-fn exec(cmd: &mut Command) -> Result<ExitStatus, Report> {
+fn exec(cmd: &mut Command) -> Result<()> {
     use color_eyre::eyre::eyre;
     use windows_sys::Win32::Foundation::{BOOL, FALSE, TRUE};
     use windows_sys::Win32::System::Console::SetConsoleCtrlHandler;
@@ -43,5 +41,10 @@ fn exec(cmd: &mut Command) -> Result<ExitStatus, Report> {
         }
     }
 
-    Ok(cmd.status()?)
+    let status = cmd.status()?;
+    if !status.success() {
+        std::process::exit(status.code().unwrap_or(0))
+    }
+
+    Ok(())
 }

--- a/wix/main.wxs
+++ b/wix/main.wxs
@@ -1,0 +1,228 @@
+<?xml version='1.0' encoding='windows-1252'?>
+<!--
+  Copyright (C) 2017 Christopher R. Field.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+<!--
+  The "cargo wix" subcommand provides a variety of predefined variables available
+  for customization of this template. The values for each variable are set at
+  installer creation time. The following variables are available:
+
+  TargetTriple      = The rustc target triple name.
+  TargetEnv         = The rustc target environment. This is typically either
+                      "msvc" or "gnu" depending on the toolchain downloaded and
+                      installed.
+  TargetVendor      = The rustc target vendor. This is typically "pc", but Rust
+                      does support other vendors, like "uwp".
+  CargoTargetBinDir = The complete path to the directory containing the
+                      binaries (exes) to include. The default would be
+                      "target\release\". If an explicit rustc target triple is
+                      used, i.e. cross-compiling, then the default path would
+                      be "target\<CARGO_TARGET>\<CARGO_PROFILE>",
+                      where "<CARGO_TARGET>" is replaced with the "CargoTarget"
+                      variable value and "<CARGO_PROFILE>" is replaced with the
+                      value from the "CargoProfile" variable. This can also
+                      be overridden manually with the "target-bin-dir" flag.
+  CargoTargetDir    = The path to the directory for the build artifacts, i.e.
+                      "target".
+  CargoProfile      = The cargo profile used to build the binaries
+                      (usually "debug" or "release").
+  Version           = The version for the installer. The default is the
+                      "Major.Minor.Fix" semantic versioning number of the Rust
+                      package.
+-->
+
+<!--
+  Please do not remove these pre-processor If-Else blocks. These are used with
+  the `cargo wix` subcommand to automatically determine the installation
+  destination for 32-bit versus 64-bit installers. Removal of these lines will
+  cause installation errors.
+-->
+<?if $(sys.BUILDARCH) = x64 or $(sys.BUILDARCH) = arm64 ?>
+    <?define PlatformProgramFilesFolder = "ProgramFiles64Folder" ?>
+<?else ?>
+    <?define PlatformProgramFilesFolder = "ProgramFilesFolder" ?>
+<?endif ?>
+
+<Wix xmlns='http://schemas.microsoft.com/wix/2006/wi'>
+
+    <Product
+        Id='*'
+        Name='beamup'
+        UpgradeCode='67FB3E2E-2204-4262-B187-C940B9E293E1'
+        Manufacturer='Tristan Sloughter'
+        Language='1033'
+        Codepage='1252'
+        Version='$(var.Version)'>
+
+        <Package Id='*'
+            Keywords='Installer'
+            Description=' Install and control active BEAM languages and components'
+            Manufacturer='Tristan Sloughter'
+            InstallerVersion='450'
+            Languages='1033'
+            Compressed='yes'
+            InstallScope='perMachine'
+            SummaryCodepage='1252'
+            />
+
+        <MajorUpgrade
+            Schedule='afterInstallInitialize'
+            DowngradeErrorMessage='A newer version of [ProductName] is already installed. Setup will now exit.'/>
+
+        <Media Id='1' Cabinet='media1.cab' EmbedCab='yes' DiskPrompt='CD-ROM #1'/>
+        <Property Id='DiskPrompt' Value='beamup Installation'/>
+
+        <Directory Id='TARGETDIR' Name='SourceDir'>
+            <Directory Id='$(var.PlatformProgramFilesFolder)' Name='PFiles'>
+                <Directory Id='APPLICATIONFOLDER' Name='beamup'>
+                    
+                    <!--
+                      Enabling the license sidecar file in the installer is a four step process:
+
+                      1. Uncomment the `Component` tag and its contents.
+                      2. Change the value for the `Source` attribute in the `File` tag to a path
+                         to the file that should be included as the license sidecar file. The path
+                         can, and probably should be, relative to this file.
+                      3. Change the value for the `Name` attribute in the `File` tag to the
+                         desired name for the file when it is installed alongside the `bin` folder
+                         in the installation directory. This can be omitted if the desired name is
+                         the same as the file name.
+                      4. Uncomment the `ComponentRef` tag with the Id attribute value of "License"
+                         further down in this file.
+                    -->
+                    <!--
+                    <Component Id='License' Guid='*'>
+                        <File Id='LicenseFile' Name='ChangeMe' DiskId='1' Source='C:\Path\To\File' KeyPath='yes'/>
+                    </Component>
+                    -->
+
+                    <Directory Id='Bin' Name='bin'>
+                        <Component Id='Path' Guid='7600FDEA-F157-41B8-A433-A81FC6A057DF' KeyPath='yes'>
+                            <Environment
+                                Id='PATH'
+                                Name='PATH'
+                                Value='[Bin]'
+                                Permanent='no'
+                                Part='last'
+                                Action='set'
+                                System='yes'/>
+                        </Component>
+                        <Component Id='binary0' Guid='*'>
+                            <File
+                                Id='exe0'
+                                Name='beamup.exe'
+                                DiskId='1'
+                                Source='$(var.CargoTargetBinDir)\beamup.exe'
+                                KeyPath='yes'/>
+                        </Component>
+                    </Directory>
+                </Directory>
+            </Directory>
+        </Directory>
+
+        <Feature
+            Id='Binaries'
+            Title='Application'
+            Description='Installs all binaries and the license.'
+            Level='1'
+            ConfigurableDirectory='APPLICATIONFOLDER'
+            AllowAdvertise='no'
+            Display='expand'
+            Absent='disallow'>
+            
+            <!--
+              Uncomment the following `ComponentRef` tag to add the license
+              sidecar file to the installer.
+            -->
+            <!--<ComponentRef Id='License'/>-->
+
+            <ComponentRef Id='binary0'/>
+
+            <Feature
+                Id='Environment'
+                Title='PATH Environment Variable'
+                Description='Add the install location of the [ProductName] executable to the PATH system environment variable. This allows the [ProductName] executable to be called from any location.'
+                Level='1'
+                Absent='allow'>
+                <ComponentRef Id='Path'/>
+            </Feature>
+        </Feature>
+
+        <SetProperty Id='ARPINSTALLLOCATION' Value='[APPLICATIONFOLDER]' After='CostFinalize'/>
+
+        
+        <!--
+          Uncomment the following `Icon` and `Property` tags to change the product icon.
+
+          The product icon is the graphic that appears in the Add/Remove
+          Programs control panel for the application.
+        -->
+        <!--<Icon Id='ProductICO' SourceFile='wix\Product.ico'/>-->
+        <!--<Property Id='ARPPRODUCTICON' Value='ProductICO' />-->
+
+        <Property Id='ARPHELPLINK' Value='https://github.com/tsloughter/beamup'/>
+        
+        <UI>
+            <UIRef Id='WixUI_FeatureTree'/>
+            
+            <!--
+              Enabling the EULA dialog in the installer is a three step process:
+
+                1. Comment out or remove the two `Publish` tags that follow the
+                   `WixVariable` tag.
+                2. Uncomment the `<WixVariable Id='WixUILicenseRtf' Value='Path\to\Eula.rft'>` tag further down
+                3. Replace the `Value` attribute of the `WixVariable` tag with
+                   the path to a RTF file that will be used as the EULA and
+                   displayed in the license agreement dialog.
+            -->
+            <Publish Dialog='WelcomeDlg' Control='Next' Event='NewDialog' Value='CustomizeDlg' Order='99'>1</Publish>
+            <Publish Dialog='CustomizeDlg' Control='Back' Event='NewDialog' Value='WelcomeDlg' Order='99'>1</Publish>
+
+        </UI>
+
+        
+        <!--
+          Enabling the EULA dialog in the installer requires uncommenting
+          the following `WixUILicenseRTF` tag and changing the `Value`
+          attribute.
+        -->
+        <!-- <WixVariable Id='WixUILicenseRtf' Value='Relative\Path\to\Eula.rtf'/> -->
+
+        
+        <!--
+          Uncomment the next `WixVariable` tag to customize the installer's
+          Graphical User Interface (GUI) and add a custom banner image across
+          the top of each screen. See the WiX Toolset documentation for details
+          about customization.
+
+          The banner BMP dimensions are 493 x 58 pixels.
+        -->
+        <!--<WixVariable Id='WixUIBannerBmp' Value='wix\Banner.bmp'/>-->
+
+        
+        <!--
+          Uncomment the next `WixVariable` tag to customize the installer's
+          Graphical User Interface (GUI) and add a custom image to the first
+          dialog, or screen. See the WiX Toolset documentation for details about
+          customization.
+
+          The dialog BMP dimensions are 493 x 312 pixels.
+        -->
+        <!--<WixVariable Id='WixUIDialogBmp' Value='wix\Dialog.bmp'/>-->
+
+    </Product>
+
+</Wix>

